apps/survey/contracts/Survey.sol

```diff
@@ -98,84 +99,104 @@ contract Survey is AragonApp {
+        }
+
+        // compute previously casted votes (i.e. substract non-used tokens from stake)
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        uint256 previousParticipation = voterStake.sub(previousVote.castedVotes[0].stake);
+        // and remove it from total participation
+        survey.participation = survey.participation.sub(previousParticipation);
+
+        // reset previously voted options, if any
+        delete survey.votes[msg.sender];
+    }
+
     /**
     * @notice Vote for multiple options in survey #`_surveyId`.
-    * @dev Empty options and stakes arrays allow to remove previous votes
     * @param _surveyId Id for survey
-    * @param _options Array with indexes of supported options
+    * @param _optionIds Array with indexes of supported options
     * @param _stakes Number of tokens assigned to each option
     */
-    function voteOptions(uint256 _surveyId, uint256[] _options, uint256[] _stakes) public {
-        require(_options.length == _stakes.length);
-        require(canVote(_surveyId, msg.sender));
+    function voteOptions(uint256 _surveyId, uint256[] _optionIds, uint256[] _stakes) public {
+        require(_optionIds.length == _stakes.length && _optionIds.length > 0);
 
         SurveyStruct storage survey = surveys[_surveyId];
 
-        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
-        MultiOptionVote storage previouslyVoted = survey.voters[msg.sender];
-
-        // revert previous votes, if any
-        if (previouslyVoted.optionsCastedLength > 0) {
-            // voter removes their vote
-            for (uint256 i = 1; i <= previouslyVoted.optionsCastedLength; i++) {
-                survey.votes[previouslyVoted.castedVotes[i].optionId] = survey.votes[previouslyVoted.castedVotes[i].optionId].sub(previouslyVoted.castedVotes[i].stake);
-            }
-            // remove previous vote participation
-            survey.participation = survey.participation.sub(voterStake.sub(previouslyVoted.castedVotes[0].stake));
-
-            // reset previously voted options, if any
-            delete survey.voters[msg.sender];
-        }
+        // revert previous votes, if any (also checks if canVote)
+        resetVote(_surveyId);
 
         uint256 totalVoted = 0;
-        // reserve first index for NO_VOTE
-        survey.voters[msg.sender].castedVotes[0] = OptionCast({optionId: NO_VOTE, stake: 0});
-        for (uint256 j = 0; j < _options.length; j++) {
-            require(_options[j] != NO_VOTE && _options[j] <= survey.options);
-            require(_stakes[j] > 0);
+        // reserve first index for ABSTAIN_VOTE
+        survey.votes[msg.sender].castedVotes[0] = OptionCast({optionId: ABSTAIN_VOTE, stake: 0});
+        for (uint256 j = 1; j <= _optionIds.length; j++) {
+            // parameter arrays don't have abstain vote, while survey ones do, so we subtract 1 here to the index
```

Comment:
Had to read this a few times to make sense of it, might be more clear being explicit: `Voters don't specify that they're abstaining, but we still keep track of this by reserving the first index of a survey's votes. We subtract 1 from the indexes of the arrays passed in by the voter to account for this.`

Maybe `j` might also be more descriptive as `optionIndex`?